### PR TITLE
feat: Allow comparing any two revisions in document history

### PR DESCRIPTION
**Pull request by RHawkins-Fisher**
Discussed in \#11981

When "Highlight changes" is enabled in the history sidebar, this adds a "Compare to" dropdown that lets you pick any revision to diff against, not just the immediately previous one. Defaults to "Previous revision" so existing behavior is unchanged.

The selected comparison target is stored in a `compareTo` query parameter, so you can share or bookmark a specific comparison (e.g. `?changes=true&compareTo=<revisionId>`). DataLoader handles fetching the comparison revision if it's not already loaded.

No API or server changes, everything is client-side. The dropdown reuses the existing InputSelect component and resets when you navigate to a different revision in the list.

### CODE REVIEW

File: app/scenes/Document/components/History.tsx [hunk 1]
```diff
@@ -38,12 +45,20 @@ function History() {
...
+  const query = useQuery();
   const sidebarContext = useLocationSidebarContext();
   const document = documents.get(match.params.documentSlug);
   const [revisionsOffset, setRevisionsOffset] = React.useState(0);
   const [eventsOffset, setEventsOffset] = React.useState(0);
   const isMobile = useMobile();
+  const userLocale = useUserLocale();
+  const [compareTo, setCompareTo] = React.useState(
+    () => query.get("compareTo") ?? COMPARE_TO_PREVIOUS
+  );
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/History.tsx:62]: No fix needed. `showChanges` in the same component follows the same pattern (initialized from URL, maintained as state). I don't see any issues with back/forward nav in practice.

File: app/scenes/Document/components/DataLoader.tsx [hunk 2]
```diff
@@ -111,25 +111,38 @@ function DataLoader({ match, children }: Props) {
...
+
+  React.useEffect(() => {
+    if (compareTo) {
+      void fetchRevisionById(compareTo, (err) =>
+        Logger.error("Failed to fetch compareTo revision", err)
+      );
     }
-    void fetchRevision();
-  }, [revisions, revisionId, document?.id]);
+  }, [fetchRevisionById, compareTo]);
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/DataLoader.tsx:145]: "previous" will never end up in the URL. The issue with `latest-<documentId>` is fixed for another comment.

File: app/scenes/Document/components/History.tsx [hunk 3]
```diff
@@ -178,6 +221,47 @@ function History() {
...
+        { locale }
+      );
+      const collaboratorName =
+        rev.collaborators?.[0]?.name ?? rev.createdBy?.name;
+
+      options.push({
+        type: "item",
+        label: dateLabel,
+        value: rev.id,
+        description: collaboratorName
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/History.tsx:258]: Fixed

File: app/scenes/Document/components/History.tsx [hunk 4]
```diff
@@ -178,6 +221,47 @@ function History() {
...
+          ? t("{{ userName }} edited", { userName: collaboratorName })
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/History.tsx]: Fixed

File: app/scenes/Document/components/RevisionViewer.tsx [hunk 5]
```diff
@@ -38,8 +40,23 @@ type Props = Omit<EditorProps, "extensions"> & {
...
+    if (!compareToParam) {
+      return undefined;
+    }
+    const resolvedId =
+      compareToParam === "latest"
+        ? RevisionHelper.latestId(revision.documentId)
+        : compareToParam;
+    return revisions.get(resolvedId);
+  }, [compareToParam, revision.documentId, revisions]);
+
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/RevisionViewer.tsx]: Fixed

File: app/scenes/Document/components/RevisionViewer.tsx [hunk 6]
```diff
@@ -38,8 +40,25 @@ type Props = Omit<EditorProps, "extensions"> & {
...
+    return compareToParam === "latest"
+      ? RevisionHelper.latestId(revision.documentId)
+      : compareToParam;
+  }, [compareToParam, revision.documentId]);
+
+  const compareToRevision = compareToRevisionId
+    ? revisions.get(compareToRevisionId)
+    : undefined;
+
+  const comparisonData = (compareToRevision ?? revision.before)?.data;
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/RevisionViewer.tsx]: Fixed

File: app/scenes/Document/components/History.tsx [hunk 7]
```diff
@@ -80,9 +95,37 @@ function History() {
...
+  // Reset "Compare to" when the user clicks a different revision in the list,
+  // but not on initial mount (which would break deep links with ?compareTo=…)
+  const prevSelectedRef = React.useRef(selectedRevisionId);
+  React.useEffect(() => {
+    if (prevSelectedRef.current !== selectedRevisionId) {
+      prevSelectedRef.current = selectedRevisionId;
+      setCompareTo(COMPARE_TO_PREVIOUS);
+      updateLocation({ compareTo: null });
+    }
+  }, [selectedRevisionId, updateLocation]);
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/History.tsx:119]: Deep linking works correctly (the `prevSelectedRef` guard skips the initial mount). The back/forward edge case is the same pattern that `showChanges` uses in this component. I'd rather keep it simple and add POP-awareness later if a maintainer asks for it.

File: app/scenes/Document/components/History.tsx [hunk 8]
```diff
@@ -178,6 +221,49 @@ function History() {
...
+    for (const rev of revisionItems) {
+      if (rev.id === resolvedSelectedId) {
+        continue;
+      }
+
+      const dateLabel = formatDate(
+        new Date(rev.createdAt),
+        "MMM do, h:mm a",
+        { locale }
+      );
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/History.tsx:250]: Month and ordinal names are localized via `dateLocale(userLocale)`. A per-locale format map (like `useLocaleTime` uses) would be more thorough but feels like scope creep for this PR. Happy to add one if a maintainer asks.

File: app/scenes/Document/components/RevisionViewer.tsx [hunk 9]
```diff
@@ -38,8 +40,27 @@ type Props = Omit<EditorProps, "extensions"> & {
...
+      : compareToParam;
+  }, [compareToParam, revision.documentId]);
+
+  const compareToRevision = compareToRevisionId
+    ? revisions.get(compareToRevisionId)
+    : undefined;
+
+  const comparisonData = compareToRevisionId
+    ? compareToRevision?.data
+    : revision.before?.data;
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/RevisionViewer.tsx:63]: In normal usage, toggling off 'Highlight changes' clears `compareTo` from the URL, so this path doesn't fire. The store lookup is O(1) even in the edge case of a manually crafted URL.

File: app/scenes/Document/components/History.tsx [hunk 10]
```diff
@@ -178,6 +221,49 @@ function History() {
...
+  const compareOptions = React.useMemo((): Option[] => {
+    const revisionItems = items.filter(
+      (item): item is Revision => item instanceof Revision
+    );
+
+    const locale = dateLocale(userLocale);
+    const resolvedSelectedId =
+      selectedRevisionId === "latest" && document
+        ? RevisionHelper.latestId(document.id)
+        : selectedRevisionId;
```
[the remark this answers is not in the document]
Comment (RHawkins-Fisher) [reply app/scenes/Document/components/History.tsx:233]: The diff itself displays correctly (DataLoader fetches the revision). The dropdown label being blank for a very old compareTo target on initial load is a minor UX gap I'm okay with for now. Scrolling the history list loads more revisions and the option appears. Can improve in a follow-up.

**Review by tommoor — APPROVED** [review]
